### PR TITLE
HTTP AutoClient NuGet package content

### LINKED ISSUE #36947: [Fundamentals] Document HTTP `AutoClient` library
Need to document the `Microsoft.Extensions.Http.AutoClient` NuGet package, make it easy to automatically create efficient and easy to use client code to invoke REST HTTP APIs.

---
[Associated WorkItem - 155413](https://dev.azure.com/msft-skilling/Content/_workitems/edit/155413)

### PR DESCRIPTION
**Pull request by IEvangelist**
## Summary

Content for the `Microsoft.Extensions.Http.AutoClient` NuGet package.

Fixes #36947

---

#### Internal previews

| 📄 File | 🔗 Preview link |
|:--|:--|
| [docs/fundamentals/networking/http/http-autoclient.md](https://github.com/dotnet/docs/blob/95652fb8e21d162c824c73fc39e05b4376de8b12/docs/fundamentals/networking/http/http-autoclient.md) | [Use the REST API HTTP client generator](https://review.learn.microsoft.com/en-us/dotnet/fundamentals/networking/http/http-autoclient?branch=pr-en-us-36970) |
| [docs/fundamentals/networking/http/httpclient.md](https://github.com/dotnet/docs/blob/95652fb8e21d162c824c73fc39e05b4376de8b12/docs/fundamentals/networking/http/httpclient.md) | [Make HTTP requests with the HttpClient](https://review.learn.microsoft.com/en-us/dotnet/fundamentals/networking/http/httpclient?branch=pr-en-us-36970) |

### CODE REVIEW

File: docs/fundamentals/networking/http/http-autoclient.md [hunk 1]
```diff
@@ -0,0 +1,290 @@
...
+    private readonly IUserClient _userClient;
+
+    public MyService(IUserClient userClient) => _userClient = userClient;
+
+    public async Task ProcessUsersAsync()
+    {
+        var users = await _myClient.GetUsersAsync();
+        // ...
+    }
+}
```
Comment (BillWagner) [inline docs/fundamentals/networking/http/http-autoclient.md:289]: Suggestion for a primary constructor:

```suggestion
public sealed class UserService(IUserClient userClient)
{
    public async Task ProcessUsersAsync()
    {
        var users = await userClient.GetUsersAsync();
        // ...
    }
}
```

File: docs/fundamentals/networking/http/http-autoclient.md [hunk 2]
```diff
@@ -0,0 +1,308 @@
...
+```
+
+In the preceding example code:
+
+- The <xref:Microsoft.Extensions.Hosting.Host.CreateApplicationBuilder(System.String[])?displayProperty=fullName> is used to create a <xref:Microsoft.Extensions.Hosting.HostApplicationBuilder>.
+- The <xref:Microsoft.Extensions.DependencyInjection.IServiceCollection> is retrieved from the <xref:Microsoft.Extensions.Hosting.HostApplicationBuilder.Services?displayProperty=nameWithType> property, to call the `AddHttpClient` extension method.
+  - The `AddHttpClient` extension method is called with the name of the <xref:System.Net.Http.HttpClient> to be registered, and a delegate to configure the <xref:System.Net.Http.HttpClient> instance.
+- The `AddUserClient` extension method is called to register the <xref:IUserClient> interface and its implementation.
+
+Consuming the client is as simple as injecting it into your service's constructor:
```
Comment (BillWagner) [inline docs/fundamentals/networking/http/http-autoclient.md:295]: nit:

```suggestion
You can consume the client by injecting it into your service's constructor:
```

**Review by BillWagner — APPROVED** [review]
I had a couple small nits, but this is ready to go.

**Comment by valentinmarinro** [comment]
Hi @IEvangelist what is the status of Microsoft.Extensions.Http.AutoClient it will still be shipped or it is discontinued, we are already using it in a production ready application and I really love it, but if it will be discontinued I will need to find alternative solutions.

**Comment by gewarren** [comment]
@joperezr FYI

**Comment by joperezr** [comment]
Hi @valentinmarinro, thanks for your question. We decided to not ship Microsoft.Extensions.AutoClient. You should consider using [Refit](https://github.com/reactiveui/refit) library instead. The API surface is super similar so it should feel familiar and it should also be very straight forward to port your existing code.

**Comment by valentinmarinro** [comment]
@joperezr thank you for the quick answer, any reasons why, it looked so beautiful and AOT compatible

**Comment by joperezr** [comment]
Main reason was the one shared above, which is when we had a closer look on Refit and reached out to the maintainers of it, it felt like its features were very closely-aligned with AutoClient, so it didn't make sense to introduce a new library when there was an existing one that would meet the same needs. In terms of the AOT-compatibility, we also discussed with the maintainers the possibility of helping Refit become AOT compatible by swapping their use of reflection for source-gen like AutoClient instead, and they seemed to be okay with us contributing that.

**Pull request merged**